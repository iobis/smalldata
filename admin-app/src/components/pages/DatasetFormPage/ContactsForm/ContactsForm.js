import classNames from 'classnames'
import InputText from '@smalldata/dwca-lib/src/components/form/InputText'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import OceanExpertNameInput from '../../OceanExpertNameInput/OceanExpertNameInput'

export default function ContactsForm({ className, contactsTableHeader, data, onChange }) {
  const { t } = useTranslation()
  const [name, setName] = useState('')
  const [email, setEmail] = useState('')
  const [organisation, setOrganisation] = useState('')
  const [contacts, setContacts] = useState(data)

  function handleAddClick() {
    const contact = { name, email, organisation }
    const newContacts = [...contacts, contact]
    setContacts(newContacts)
    onChange(newContacts)
  }

  function handleRemoveClick(index) {
    const newContacts = contacts.filter((_, i) => i !== index)
    setContacts(newContacts)
    onChange(newContacts)
  }

  function handleOceanExpertProfileChange(profile) {
    setName(profile.name)
    setEmail(profile.email || email)
  }

  return (
    <div className={classNames('contacts-form', className)}>
      <OceanExpertNameInput
        oceanExpertName={name}
        onChange={handleOceanExpertProfileChange}/>
      <div className="columns column">
        <InputText
          className="column email"
          name="datasetFormPage.contactsForm.email"
          onChange={value => setEmail(value)}
          value={email}/>
        <InputText
          className="column organisation"
          name="datasetFormPage.contactsForm.organisation"
          onChange={value => setOrganisation(value)}
          value={organisation}/>
        <div className="column field button-placeholder">
          <button className="add button" onClick={handleAddClick}>{t('common.add')}</button>
        </div>
      </div>
      <div className="column field">
        <div className="title is-5 contacts-table-header">{contactsTableHeader}</div>
        <table className="general table is-fullwidth is-striped is-hoverable">
          <thead>
            <tr>
              <th>{t('datasetFormPage.contactsForm.name.label')}</th>
              <th>{t('datasetFormPage.contactsForm.email.label')}</th>
              <th>{t('datasetFormPage.contactsForm.organisation.label')}</th>
              <th/>
            </tr>
          </thead>
          <tbody>
            {contacts.map((contact, index) => (
              // eslint-disable-next-line react/no-array-index-key
              <ContactRow key={index} {...contact} onRemove={() => handleRemoveClick(index)}/>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  )
}

export const contactShape = {
  email:        PropTypes.string.isRequired,
  name:         PropTypes.string.isRequired,
  organisation: PropTypes.string.isRequired
}

ContactsForm.propTypes = {
  className:           PropTypes.string,
  contactsTableHeader: PropTypes.string.isRequired,
  data:                PropTypes.arrayOf(PropTypes.shape(contactShape)).isRequired,
  onChange:            PropTypes.func.isRequired
}

function ContactRow({ email, name, organisation, onRemove }) {
  const { t } = useTranslation()
  return (
    <tr className="contact-row">
      <td className="name">{name}</td>
      <td className="email">{email}</td>
      <td className="organisation">{organisation}</td>
      <td className="action">
        <button className="remove button" onClick={onRemove}>
          {t('common.remove')}
        </button>
      </td>
    </tr>
  )
}

ContactRow.propTypes = {
  email:        PropTypes.string.isRequired,
  name:         PropTypes.string.isRequired,
  onRemove:     PropTypes.func.isRequired,
  organisation: PropTypes.string.isRequired
}
