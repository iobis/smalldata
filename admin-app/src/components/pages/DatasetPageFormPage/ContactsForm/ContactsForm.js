import classNames from 'classnames'
import InputText from '@smalldata/dwca-lib/src/components/form/InputText'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'

export default function ContactsForm({ className, contactsTableHeader, data, onChange }) {
  const { t } = useTranslation()
  const [name, setName] = useState('')
  const [email, setEmail] = useState('')
  const [organisation, setOrganisation] = useState('')
  const [position, setPosition] = useState('')
  const [contacts, setContacts] = useState(data)

  function handleAddClick() {
    const contact = { name, email, organisation, position }
    const newContacts = [...contacts, contact]
    setContacts(newContacts)
    onChange(newContacts)
  }

  return (
    <div className={classNames('contacts-form', className)}>
      <InputText
        className="name"
        name="datasetPageFormPage.contactsForm.name"
        onChange={value => setName(value)}
        value={name}/>
      <InputText
        className="email"
        name="datasetPageFormPage.contactsForm.email"
        onChange={value => setEmail(value)}
        value={email}/>
      <InputText
        className="organisation"
        name="datasetPageFormPage.contactsForm.organisation"
        onChange={value => setOrganisation(value)}
        value={organisation}/>
      <InputText
        className="position"
        name="datasetPageFormPage.contactsForm.position"
        onChange={value => setPosition(value)}
        value={position}/>
      <div className="column field">
        <button className="add button" onClick={handleAddClick}>{t('common.add')}</button>
      </div>
      <div className="column field">
        <div className="title is-5 contacts-table-header">{contactsTableHeader}</div>
        <table className="general table is-fullwidth is-striped is-hoverable">
          <thead>
            <tr>
              <th>{t('datasetPageFormPage.contactsForm.name.label')}</th>
              <th>{t('datasetPageFormPage.contactsForm.email.label')}</th>
              <th>{t('datasetPageFormPage.contactsForm.organisation.label')}</th>
              <th>{t('datasetPageFormPage.contactsForm.position.label')}</th>
            </tr>
          </thead>
          <tbody>
            {contacts.map((contact, index) => (
              // eslint-disable-next-line react/no-array-index-key
              <ContactRow key={index} {...contact}/>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  )
}

export const contactShape = {
  email:        PropTypes.string.isRequired,
  name:         PropTypes.string.isRequired,
  organisation: PropTypes.string.isRequired,
  position:     PropTypes.string.isRequired
}

ContactsForm.propTypes = {
  className:           PropTypes.string,
  contactsTableHeader: PropTypes.string.isRequired,
  data:                PropTypes.arrayOf(PropTypes.shape(contactShape)).isRequired,
  onChange:            PropTypes.func.isRequired
}

function ContactRow({ email, name, organisation, position }) {
  return (
    <tr className="contact-row">
      <td className="name">{name}</td>
      <td className="email">{email}</td>
      <td className="organisation">{organisation}</td>
      <td className="position">{position}</td>
    </tr>
  )
}

ContactRow.propTypes = {
  email:        PropTypes.string.isRequired,
  name:         PropTypes.string.isRequired,
  organisation: PropTypes.string.isRequired,
  position:     PropTypes.string.isRequired
}
