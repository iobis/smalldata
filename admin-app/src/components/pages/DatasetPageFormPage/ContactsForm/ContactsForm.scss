.contacts-form {
  .button-placeholder {
    align-self: flex-end;
    margin-bottom: 0.75rem;
  }
}
