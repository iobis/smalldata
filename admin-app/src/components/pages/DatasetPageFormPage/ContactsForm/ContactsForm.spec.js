import ContactsForm from './ContactsForm'
import React from 'react'
import { mount } from 'enzyme'

describe('ContactsForm', () => {
  let onChange
  let wrapper

  it('returns contacts when adding to empty list', () => {
    onChange = jest.fn()
    wrapper = mount(
      <ContactsForm
        className="custom-class"
        contactsTableHeader="Contacts Table Header"
        data={[]}
        onChange={onChange}/>)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.contact-row')).toHaveLength(0)

    wrapper.find('.name input').simulate('change', { target: { value: 'name-1' } })
    wrapper.find('.email input').simulate('change', { target: { value: 'email-1' } })
    wrapper.find('.organisation input').simulate('change', { target: { value: 'organisation-1' } })
    wrapper.find('.add').simulate('click')
    expect(wrapper.find('.contact-row')).toHaveLength(1)
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenNthCalledWith(1, [{
      name:         'name-1',
      email:        'email-1',
      organisation: 'organisation-1'
    }])
  })

  it('returns updated contacts when adding to list with single element', () => {
    onChange = jest.fn()
    wrapper = mount(
      <ContactsForm
        className="custom-class"
        contactsTableHeader="Contacts Table Header"
        data={[{
          name:         'name-1',
          email:        'email-1',
          organisation: 'organisation-1'
        }]}
        onChange={onChange}/>)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.contact-row')).toHaveLength(1)

    wrapper.find('.name input').simulate('change', { target: { value: 'name-2' } })
    wrapper.find('.email input').simulate('change', { target: { value: 'email-2' } })
    wrapper.find('.organisation input').simulate('change', { target: { value: 'organisation-2' } })
    wrapper.find('.add').simulate('click')
    expect(wrapper.find('.contact-row')).toHaveLength(2)
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenNthCalledWith(1, [{
      name:         'name-1',
      email:        'email-1',
      organisation: 'organisation-1'
    }, {
      name:         'name-2',
      email:        'email-2',
      organisation: 'organisation-2'
    }])
  })

  describe('when removing contacts', () => {
    const data = [{
      name:         'name-1',
      email:        'email-1',
      organisation: 'organisation-1'
    }, {
      name:         'name-2',
      email:        'email-2',
      organisation: 'organisation-2'
    }]

    it('removes first contact', () => {
      onChange = jest.fn()
      wrapper = mount(
        <ContactsForm
          contactsTableHeader="Contacts Table Header"
          data={data}
          onChange={onChange}/>)
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find('.contact-row')).toHaveLength(2)
      expect(wrapper.find('.contact-row .remove')).toHaveLength(2)

      wrapper.find('.contact-row .remove').at(0).simulate('click')
      expect(wrapper.find('.contact-row')).toHaveLength(1)
      expect(onChange).toHaveBeenCalledTimes(1)
      expect(onChange).toHaveBeenNthCalledWith(1, [data[1]])
    })

    it('removes second contact', () => {
      onChange = jest.fn()
      wrapper = mount(
        <ContactsForm
          contactsTableHeader="Contacts Table Header"
          data={data}
          onChange={onChange}/>)
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find('.contact-row')).toHaveLength(2)
      expect(wrapper.find('.contact-row .remove')).toHaveLength(2)

      wrapper.find('.contact-row .remove').at(1).simulate('click')
      expect(wrapper.find('.contact-row')).toHaveLength(1)
      expect(onChange).toHaveBeenCalledTimes(1)
      expect(onChange).toHaveBeenNthCalledWith(1, [data[0]])
    })
  })
})
