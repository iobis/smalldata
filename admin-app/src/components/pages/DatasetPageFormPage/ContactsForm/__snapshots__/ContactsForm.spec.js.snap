// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsForm returns contacts when adding 1`] = `
<ContactsForm
  className="custom-class"
  contactsTableHeader="Contacts Table Header"
  data={Array []}
  onChange={[MockFunction]}
>
  <div
    className="contacts-form custom-class"
  >
    <InputText
      className="name"
      name="datasetPageFormPage.contactsForm.name"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field name"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.name.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <InputText
      className="email"
      name="datasetPageFormPage.contactsForm.email"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field email"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.email.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <InputText
      className="organisation"
      name="datasetPageFormPage.contactsForm.organisation"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field organisation"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.organisation.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <InputText
      className="position"
      name="datasetPageFormPage.contactsForm.position"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field position"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.position.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <div
      className="column field"
    >
      <button
        className="add button"
        onClick={[Function]}
      >
        common.add
      </button>
    </div>
    <div
      className="column field"
    >
      <div
        className="title is-5 contacts-table-header"
      >
        Contacts Table Header
      </div>
      <table
        className="general table is-fullwidth is-striped is-hoverable"
      >
        <thead>
          <tr>
            <th>
              datasetPageFormPage.contactsForm.name.label
            </th>
            <th>
              datasetPageFormPage.contactsForm.email.label
            </th>
            <th>
              datasetPageFormPage.contactsForm.organisation.label
            </th>
            <th>
              datasetPageFormPage.contactsForm.position.label
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</ContactsForm>
`;

exports[`ContactsForm returns updated contacts when when adding 1`] = `
<ContactsForm
  className="custom-class"
  contactsTableHeader="Contacts Table Header"
  data={
    Array [
      Object {
        "email": "email-1",
        "name": "name-1",
        "organisation": "organisation-1",
        "position": "position-1",
      },
    ]
  }
  onChange={[MockFunction]}
>
  <div
    className="contacts-form custom-class"
  >
    <InputText
      className="name"
      name="datasetPageFormPage.contactsForm.name"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field name"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.name.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <InputText
      className="email"
      name="datasetPageFormPage.contactsForm.email"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field email"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.email.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <InputText
      className="organisation"
      name="datasetPageFormPage.contactsForm.organisation"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field organisation"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.organisation.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <InputText
      className="position"
      name="datasetPageFormPage.contactsForm.position"
      onChange={[Function]}
      value=""
    >
      <div
        className="column field position"
      >
        <label
          className="label"
        >
          datasetPageFormPage.contactsForm.position.label
        </label>
        <input
          className="input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </InputText>
    <div
      className="column field"
    >
      <button
        className="add button"
        onClick={[Function]}
      >
        common.add
      </button>
    </div>
    <div
      className="column field"
    >
      <div
        className="title is-5 contacts-table-header"
      >
        Contacts Table Header
      </div>
      <table
        className="general table is-fullwidth is-striped is-hoverable"
      >
        <thead>
          <tr>
            <th>
              datasetPageFormPage.contactsForm.name.label
            </th>
            <th>
              datasetPageFormPage.contactsForm.email.label
            </th>
            <th>
              datasetPageFormPage.contactsForm.organisation.label
            </th>
            <th>
              datasetPageFormPage.contactsForm.position.label
            </th>
          </tr>
        </thead>
        <tbody>
          <ContactRow
            email="email-1"
            key="0"
            name="name-1"
            organisation="organisation-1"
            position="position-1"
          >
            <tr
              className="contact-row"
            >
              <td
                className="name"
              >
                name-1
              </td>
              <td
                className="email"
              >
                email-1
              </td>
              <td
                className="organisation"
              >
                organisation-1
              </td>
              <td
                className="position"
              >
                position-1
              </td>
            </tr>
          </ContactRow>
        </tbody>
      </table>
    </div>
  </div>
</ContactsForm>
`;
